stan/math/rev/mat/functor/algebra_solver.hpp

```diff
@@ -0,0 +1,246 @@
+
+        theta_[0] = this;
+        for (int i = 1; i < x.size(); ++i)
+          theta_[i] = new vari(theta_dbl(i), false);
+
+        // Compute the Jacobian
+        Eigen::MatrixXd Jf_x = fx.get_jacobian(theta_dbl);
+        hybrj_functor_solver<FS, F, double, double>
+          fy(fs, f, theta_dbl, value_of(y), dat, dat_int, msgs, false);
+        Eigen::MatrixXd Jf_y = fy.get_jacobian(value_of(y));
+
+        Jx_y_ = - stan::math::mdivide_left(Jf_x, Jf_y);
+      }
+
+      void chain() {
+        for (int i = 0; i < x_size_; i++)
+          for (int j = 0; j < y_size_; j++)
+            y_[j]->adj_ += theta_[i]->adj_ * Jx_y_(i, j);
+      }
+    };
+
+    /**
+     * Return the solution to the specified system of algebraic
+     * equations given an initial guess, and parameters and data,
+     * which get passed into the algebraic system. The user can
+     * also specify the relative tolerance (xtol in Eigen's code),
+     * the function tolerance, and the maximum number of steps
+     * (maxfev in Eigen's code).
+     *
+     * Throw an exception if the norm of f(x), where f is the
+     * output of the algebraic system and x the proposed solution,
+     * is greater than the function tolerance. We here use the
+     * norm as a metric to measure how far we are from 0.
+     *
+     * @tparam F type of equation system function.
+     * @param[in] f Functor that evaluates the system of equations.
+     * @param[in] x Vector of starting values.
+     * @param[in] y parameter vector for the equation system. The function
+     *            is overloaded to treat y as a vector of doubles or of a
+     *            a template type T.
+     * @param[in] dat continuous data vector for the equation system.
+     * @param[in] dat_int integer data vector for the equation system.
+     * @param[in, out] msgs the print stream for warning messages.
+     * @param[in] relative_tolerance determines the convergence criteria
+     *            for the solution.
+     * @param[in] function_tolerance determines whether roots are acceptable.
+     * @param[in] max_num_steps  maximum number of function evaluations.
+     * @return theta Vector of solutions to the system of equations.
+     */
+    template <typename F>
+    Eigen::VectorXd
+    algebra_solver(const F& f,
+                   const Eigen::VectorXd& x,
+                   const Eigen::VectorXd& y,
+                   const std::vector<double>& dat,
+                   const std::vector<int>& dat_int,
+                   std::ostream* msgs = 0,
+                   double relative_tolerance = 1e-10,
+                   double function_tolerance = 1e-6,
+                   long int max_num_steps = 1e+3) {  // NOLINT(runtime/int)
+      // Check that arguments are valid
+      check_nonzero_size("algebra_solver", "initial guess", x);
+      check_nonzero_size("algebra_solver", "parameter vector", y);
+      for (int i = 0; i < x.size(); i++)  // FIX ME - do these w/o for loop?
```

Comment:
There is a vectorized version in `check_finite.hpp` but it doesn't seem to work on an `Eigen::MatrixXd`. See error at compilation:
```
./stan/math/prim/scal/fun/value_of_rec.hpp:24:14: error: cannot convert 'const Eigen::Matrix<double, -1,
      1, 0, -1, 1>' to 'double' without a conversion operator
      return static_cast<double>(x);
             ^~~~~~~~~~~~~~~~~~~~~~
./stan/math/prim/scal/err/check_finite.hpp:33:42: note: in instantiation of function template
      specialization 'stan::math::value_of_rec<Eigen::Matrix<double, -1, 1, 0, -1, 1> >' requested here
            if (!(boost::math::isfinite)(value_of_rec(stan::get(y, n))))
                                         ^
./stan/math/prim/scal/err/check_finite.hpp:61:11: note: in instantiation of member function
      'stan::math::(anonymous namespace)::finite<Eigen::Matrix<double, -1, 1, 0, -1, 1>, true>::check'
      requested here
        ::check(function, name, y);
          ^
./stan/math/rev/mat/functor/algebra_solver.hpp:119:7: note: in instantiation of function template
      specialization 'stan::math::check_finite<Eigen::Matrix<double, -1, 1, 0, -1, 1> >' requested here
      check_finite("algebra_solver", "initial guess", x);
      ^
test/unit/math/rev/mat/functor/algebra_solver_test.cpp:88:10: note: in instantiation of function
      template specialization 'stan::math::algebra_solver<simpleEq_functor>' requested here
 theta = algebra_solver(simpleEq_functor(),
         ^
1 error generated.
```